sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/MethodParameterLookupTest.cs

```diff
@@ -0,0 +1,262 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.UnitTest.TestFramework;
+using CSharpSyntax = Microsoft.CodeAnalysis.CSharp.Syntax;
+using VBSyntax = Microsoft.CodeAnalysis.VisualBasic.Syntax;
+
+namespace SonarAnalyzer.UnitTest.Helpers
+{
+    [TestClass]
+    public class MethodParameterLookupTest
+    {
+
+        [TestMethod]
+        public void TestMethodParameterLookup_CS()
+        {
+            const string Source = @"
+namespace Test
+{
+    class TestClass
+    {
+        void Main()
+        {
+            Special(65535);
+            DoNothing();
+            DoSomething(42, true, null);
+            DoSomething(42, true, null, ""Ipsum"");
+            DoSomething(b: true, a: 42, c: null);
+            DoSomething(b: true, a: 42, c: null, d: ""Ipsum"");
+        }
+
+        void DoSomething(int a, bool b, object c, string d = ""Lorem"")
+        {
+            if (c != null)
+            {
+                Console.WriteLine($""{a}#: {b}: {d}"");
+            }
+        }
+
+        void DoNothing() 
+        {
+            Console.WriteLine(""Nothing"");
+        }
+
+        void Special(int special){
+        }
+
+    }
+}";
+            var compiler = new SnippetCompiler(Source, ignoreErrors: true, language: AnalyzerLanguage.CSharp);
+            var mainMethod = compiler.GetNodes<CSharpSyntax.MethodDeclarationSyntax>().Single(x => x.Identifier.ValueText == "Main");
+            var specialInvocation = compiler.GetNodes<CSharpSyntax.InvocationExpressionSyntax>().Single(x => x.Expression is CSharpSyntax.IdentifierNameSyntax identifier && identifier.Identifier.ValueText == "Special");
+            var specialArgument = specialInvocation.ArgumentList.Arguments.Single();
+            var specialParameter = (compiler.SemanticModel.GetSymbolInfo(specialInvocation).Symbol as IMethodSymbol).Parameters.Single();
+            Assert.IsNotNull(mainMethod);   
+
+            foreach (var invocation in mainMethod.DescendantNodes().OfType<CSharpSyntax.InvocationExpressionSyntax>())
+            {
+                var method = compiler.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+                var lookup = new CSharpMethodParameterLookup(invocation.ArgumentList, method);
+                var context = new CSharpInspectionContext();
+                context.Init(compiler, lookup, method, invocation.ArgumentList.Arguments.ToArray(), specialArgument, specialParameter);
+                InspectLookupValues(context);
+            }
```

Comment:
It is really hard to read and understand what the test actually does, as it has logic regarding expected values spread across several helper methods, and the for-loop makes it hard to know what we are testing at one point.
It would be easier to have one helper method doing the all the logic and one doing the assertions on what we expect.

I would have a much easier time reading something similar to:
```
var doSomethingFirstInvocation = GetInvocationExpression(mainMethod, "DoSomething")[0];
assertParameterNames(doSomethingFirstInvocation, "a", "b", "c");
assertParameterValues(doSomethingFirstInvocation, 42, true, null, "Lorem");

var doSomethingSecondInvocation = GetInvocationExpression(mainMethod, "DoSomething")[1];
assertParameterNames(doSomethingSecondInvocation , "a", "b", "c");
assertParameterValues(doSomethingSecondInvocation , 42, true, null, "Lorem");
```

Which would also allow to test different values for the parameters, as at the moment each parameter has a static mapping between parameter name and parameter value.